orttraining/orttraining/core/optimizer/megatron_transformer.cc

```diff
@@ -150,44 +171,105 @@ bool MegatronTransformer::PartitionWeightByColumn(const Graph& graph, const Node
 
-  std::string new_initializer_name = input_arg.Name() + "_column_rank_" + std::to_string(horizontal_parallel_rank_);
+  std::string new_initializer_name = original_name + "_column_rank_" + std::to_string(horizontal_parallel_rank_);
 
   initializer_partition.set_name(new_initializer_name);
   initializer_partition.set_data_type(data_type);
 
   int64_t column_partition = column_count / horizontal_parallel_size_;
   int64_t column_stride = column_count / stride;
-  int64_t column_stride_partition = column_stride / horizontal_parallel_size_;
 
+  std::vector<int64_t> new_shape;
   if (rank == 2) {
     initializer_partition.add_dims(row_count);
+    new_shape.push_back(row_count);
   }
 
   initializer_partition.add_dims(column_partition);
+  new_shape.push_back(column_partition);
   const int64_t element_count = row_count * column_partition;
 
   std::vector<float> result;
   result.reserve(element_count);
 
-  const int64_t stride_partition_column_offset = horizontal_parallel_rank_ * column_stride_partition;
-  for (auto row_index = 0; row_index < row_count; row_index++) {
-    auto row_offset = row_index * column_count;
-    for (auto stride_index = 0; stride_index < stride; stride_index++) {
-      auto column_offset = row_offset + stride_index * column_stride + stride_partition_column_offset;
-      std::copy(a_weight + column_offset, a_weight + column_offset + column_stride_partition, std::back_inserter(result));
+  PartitionBufferByColumn(a_weight, row_count, column_count, column_stride, stride, result);
+  initializer_partition.set_raw_data(result.data(), element_count * sizeof(float));
+
+  // Partition initial optimizer state if available
+  const auto optim_state_it = initial_optimizer_states_.find(original_name);
+  if (optim_state_it != initial_optimizer_states_.end()) {
+    auto& initial_states = optim_state_it->second;
+    // partition moments same way as the weight
+    for (const auto& moments_prefix : training::MOMENTS_PREFIXES) {
+      const auto initial_state_it = initial_states.find(moments_prefix);
+      if (initial_state_it != initial_states.end()) {
+        auto* init_tensor = initial_state_it->second.GetMutable<Tensor>();
+
+        OrtValue partitioned;
+        auto element_type = init_tensor->DataType();
+        TensorShape partition_shape(new_shape);
+        std::unique_ptr<Tensor> p_tensor;
+
+        if (utils::IsPrimitiveDataType<float>(element_type)) {
+          float* data_buffer = init_tensor->MutableData<float>();
+
+          // allocate temporary memory to get the column partitioned state
+          std::vector<float> result_buffer;
+          result_buffer.reserve(element_count);
+          PartitionBufferByColumn(data_buffer, row_count, column_count, column_stride, stride, result_buffer);
+
+          // allocate a new buffer to hold the partitioned optimizer state 
+          // as column partitioning cannot re-use the original
+          // buffer as its non-contiguous read
```

Comment:
sorry I am still not clear about this part. why cannot we allocate a new ONNX_NAMESPACE::TensorProto just as we did for newly created partition weight, instead of using CPU EP allocator to create a buffer?

Reply:
as discussed offline, we need to currently maintain the initial optimizer states as an OrtValue, which is converted eventually to a TensorProto in the optimizer builder after Megatron and Zero partitioning. This approach saves CPU memory as creating a TensorProto involves a copy, and by delaying the copy until after the partitioning results in a smaller copy only for the optimizer states currently present on the rank.